### PR TITLE
[998] Add providers endpoint and serializer

### PR DESCRIPTION
**Pull request by dankmitchell**
### Context
V2 Providers endpoint

### Changes proposed in this pull request
Add v2 providers serialiser 

### Guidance to review
`api/v2/users/:user_id/providers`

### CODE REVIEW

**Comment by dankmitchell** [comment]
[Add api/v2/providers endpoint](https://trello.com/c/erPNdfLH/998-add-api-v2-providers-endpoint)

File: config/routes.rb [hunk 1]
```diff
@@ -39,6 +39,7 @@
 
     namespace :v2 do
       resources :users
+      resources :providers
       scope :providers do
```
Comment (defong) [inline config/routes.rb]: Can this be nested inside of the `resource` one?

Comment (misaka) [reply config/routes.rb]: I think @defong is right ... even though I think we should have `/providers`, I think that for the sake of the API we are actually requesting the user's list of providers, which means using `/users/:user_id/providers` to make that explicit.

File: app/controllers/api/v2/application_controller.rb [hunk 2]
```diff
@@ -4,6 +4,7 @@ class ApplicationController < ::ApplicationController
       SERIALIZABLE_CLASSES = {
         User: API::V2::UserSerializable,
         Course: API::V2::CourseSerializable,
+        Provider: API::V2::ProviderSerializable
```
Comment (defong) [inline app/controllers/api/v2/application_controller.rb]: Just wondering 
image

http://jsonapi-rb.org/guides/getting_started/rails.html 

does it mean we can omit it?

Comment (dankmitchell) [reply app/controllers/api/v2/application_controller.rb]: Doing this in a separate PR @defong

File: app/policies/provider_policy.rb [hunk 3]
```diff
@@ -20,6 +20,6 @@ def initialize(user, provider)
   end
 
   def index?
-    user.providers.include?(provider)
```
Comment (dankmitchell) [inline app/policies/provider_policy.rb]: @misaka this change is breaking the `/providers/:id/courses` request spec...

**Comment by dankmitchell** [comment]
@misaka @benilovj i've been thinking about this and wondering if this PR should be changed to just `/providers` and then use the pundit scope to decide which providers the authenticated user can view? In the same way the `providers/:id/courses` endpoint is setup...

From my limited understanding of pundit we can easily add a scope to provider policy to allow an "admin" to view all the providers...

File: app/policies/provider_policy.rb [hunk 4]
```diff
@@ -19,7 +19,14 @@ def initialize(user, provider)
...
+  def can_list_providers?
+    user.present?
+  end
+
+  def can_manage_provider?
     user.providers.include?(provider)
   end
+
+  alias_method :index?, :can_list_providers?
+  alias_method :show?, :can_manage_provider?
```
Comment (misaka) [inline app/policies/provider_policy.rb]: My first intuition was to do this like so:

```
def index?
  user.present?
end

def show?
   user.providers.include?(provider)
end

alias_method :can_list_courses?, :show?
```

Comment (benilovj) [reply app/policies/provider_policy.rb]: Fixed.

**Comment by benilovj** [comment]
@misaka ready for another review 😅

File: spec/policies/user_policy_spec.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+require "rails_helper"
+
+describe UserPolicy do
+  describe 'show?' do
```
Comment (misaka) [inline spec/policies/user_policy_spec.rb]: Backfilling tests. :100:

**Comment by benilovj** [comment]
> Sorry to pick but there's one new method that's not covered.
> 
> I can add the test for it though.

If you could add a spec, that'd help get this across the line 👍 thanks

**Comment by misaka** [comment]
> > Sorry to pick but there's one new method that's not covered.
> > I can add the test for it though.
> 
> If you could add a spec, that'd help get this across the line 👍 thanks

Done. LGTM!

**Pull request merged**